rust/alloc/string.rs

```diff
@@ -2845,3 +3004,107 @@ impl From<char> for String {
         c.to_string()
     }
 }
+
+/// A helper type implementing non-panicking version of [`Write`] trait for [`String`].
+///
+/// The `StringWriter` type is a helper type, that implements [`Write`] trait
+/// for a [`String`] type in a way that never panics during memory allocation.
+/// Instead, if any memory allocation during all writing and formatting
+/// operations executed on the `StringWriter` fail, it saves the error and
+/// returns it as a result when the result of the writing and formatting
+/// is requested.
+///
+/// [`Write`]: fmt::Write
+///
+/// # Panics
+///
+/// `Write` trait implementation for `StringWriter` panics if a formatting
+/// trait implementation returns an error. This indicates an incorrect
+/// implementation since `fmt::Write for String` never returns an error itself.
+///
+/// # Examples
+///
+/// Basic usage:
+/// ```
+/// use alloc::collections::TryReserveError;
+/// use core::fmt::Write;
+///
+/// fn write_hello_world() -> Result<String, TryReserveError> {
+///     let mut writer = StringWriter::try_with_capacity(0)?;
+///     writer.write_fmt(format_args!("Hello, {}!", "world")).unwrap();
+///     writer.result()
+/// }
+///
+/// assert_eq!(write_hello_world().unwrap(), "Hello, World!");
+/// ```
+#[stable(feature = "kernel", since = "1.0.0")]
+#[derive(Debug)]
+pub struct StringWriter {
```

Comment:
Personally, I do not like the name `StringWriter`. The information about the panic-free logic ant it's purpose is missing, but I had no better idea for the name to be descriptive but also short and clean. If you have any ideas for a better name, feel free to comment.

Reply:
This shouldn't be `pub`.

Reply:
I have made it public because it seemed to me that it may be useful in some other places too. For example, one may use it like that:
```
let mut sw = StringWriter::try_with_capacity(0)?;
write!(&mut sw, "test {}\n", 1).unwrap();
write!(&mut sw, "test {}\n", 2).unwrap();
let s = sw.into_string()?;
```
What do you think?